light-client/src/supervisor.rs

```diff
@@ -15,6 +15,39 @@ use crate::{
     types::{Height, LightBlock, PeerId, Status},
 };
 
+pub trait Handle {
+    /// Get latest trusted block from the [`Supervisor`].
+    fn latest_trusted(&mut self) -> Result<Option<LightBlock>, Error>;
+
+    /// Verify to the highest block.
+    fn verify_to_highest(&mut self) -> Result<LightBlock, Error>;
+
+    /// Verify to the block at the given height.
+    fn verify_to_target(&mut self, height: Height) -> Result<LightBlock, Error>;
+
+    /// Async version of `verify_to_highest`.
+    ///
+    /// The given `callback` will be called asynchronously with the
+    /// verification result.
+    fn verify_to_highest_async(
+        &mut self,
+        callback: impl FnOnce(Result<LightBlock, Error>) -> () + Send + 'static,
+    );
+
+    /// Async version of `verify_to_target`.
+    ///
+    /// The given `callback` will be called asynchronously with the
+    /// verification result.
+    fn verify_to_target_async(
+        &mut self,
+        height: Height,
+        callback: impl FnOnce(Result<LightBlock, Error>) -> () + Send + 'static,
+    );
+
+    /// Terminate the underlying [`Supervisor`].
+    fn terminate(&mut self);
+}
```

Comment:
Looks good to me 👍 My understanding of why making this a trait is cool is because it makes testability easier. Is there a (wip) example of such a test that would be annoying to write without this?

Reply:
The commit https://github.com/informalsystems/tendermint-rs/pull/363/commits/a452244ab2ab4802bb6dd4a85a3c87aec5e5f453 shows the start, where the test code implements a mockHandle. Generally it means a reduction of the objects that need constructing. If something was dependent on the `Handle` before, one had to provide a `PeerList`, `ForkDetector`, `EvidenceProvider` to construct the `Supervisor`. On all the state inside had to be correct in order to produce output as expected, which means the correct state in peer(s) stores, etc.

Reply:
Thanks!